sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -930,6 +930,18 @@ class CodegenContext {
       // inline execution if only one block
       blocks.head
     } else {
+      if (Utils.isTesting) {
+        // Passing global variables to the split method is dangerous, as any mutating to it is
+        // ignored and may lead to unexpected behavior.
+        // We don't need to check `arrayCompactedMutableStates` here, as it results to array access
```

Comment:
what if we declare a variable with the same name of an `arrayCompactedMutableStates `? Let's say that we have:
```
public class Foo {
 private Object[] ourArray;
 // ....
 private void ourMethod() {
   Object[] ourArray = new Object[1];
   ourSplitFunction(ourArray);
 }
 private void ourSplitFunction(Object[] ourArray) {
  ourArray[0] = null;
 }
```

Reply:
is there any place we would do that? `ctx.addMutableState` returns an array access code, I can't image a caller would extract the array name from it and use it as parameters...

Reply:
I don't think so, but currently there is also no place which creates the problem for which this assertion is being introduces. Of course this case is very very unlikely, but since we are introducing the check, I think that the effort to ensure also this very remote corner case is very low...

Reply:
I think it is no place now, but there could be (a place)[https://github.com/apache/spark/pull/19865#issuecomment-348777024]. If `smj_value16` is replaced with an array element by `ctx.addMutableState`, the array element was passed.

(This PR)[https://github.com/apache/spark/pull/19937] fixed this issue by eliminating usage of mutable state.